pkgs/applications/networking/cluster/k3s/docs/PKG_UPKEEP.md

```diff
@@ -0,0 +1,60 @@
+
+# K3s Upkeep for Maintainers
+
+General documentation for the K3s maintainer and reviewer use for consistency in maintenance processes.
+
+## NixOS Release Maintenance
+
+This process split into two sections and adheres to the versioning policy outlined in [VERSIONING.md](VERSIONING.md).
+
+### Pre-Release
+
+* Prior to the breaking change window of the next release being closed:
+  * `nixos-unstable`: Ensure k3s points to latest versioned release
+  * `nixos-unstable`: Ensure release notes are up to date
+  * `nixos-unstable`: Remove k3s releases which will be end of life upstream prior to end-of-life for the next NixOS stable release are removed with proper deprecation notice (process listed below)
+
+### Post-Release
+
+* For major/minor releases of k3s:
+  * `nixos-unstable`: Create a new versioned k3s package
+  * `nixos-unstable`: Update k3s alias to point to new versioned k3s package
+  * `nixos-unstable`: Add NixOS Release note denoting:
+    * Removal of deprecated K3s packages
+    * Migration information from the Kubernetes and K3s projects
+  * `nixos-stable`: Backport the versioned package
+* For patch releases of existing packages:
+  * `nixos-unstable`: Update package version (process listed below)
+  * `nixos-stable`: Backport package update done to nixos-unstable
+
+## Patch Upgrade Process
+
+Patch upgrades can use the [update script](../update-script.sh) in the root of the package. To update k3s 1.30.x, for example, you can run the following from the root of the nixpkgs git repo:
+
+> ./pkgs/applications/networking/cluster/k3s/update-script.sh "30"
+
+To update another version, just replace the `"30"` with the appropriate minor revision.
+
+If the script should fail, the first goal would be to fix the script. If you are unable to fix the script, open an issue reporting the update script failure with the exact command used and the failure observed.
+
+RyanTM bot can automatically do patch upgrades. Update logs are available at versioned urls, e.g. for 1.30.x: https://r.ryantm.com/log/k3s_1_30
+
+## Package Removal Process
+
```

Comment:
Packages should be removed from `nixos-unstable` whenever goes [EOL at upstream](https://kubernetes.io/releases/patch-releases/#non-active-branch-history).

Reply:
I was keeping this file the "How" and have documented the "Why" in VERSIONING.md, but I have a thought on resolving this which I can commit in a few hours.

Reply:
@wrmilling Do you want to add the dates in advance like that?
Let me know when you want the PR to be merged.

Reply:
I think it's worth linking the upstream EOL page (https://endoflife.date/kubernetes) and mentioning that we can consult that.

Since we can link to that, I think it's nicer to avoid duplicating the table here, since it'll just get stale.

Reply:
Linked to the relevant section in versioning.md which includes the upstream documentation (which has maintenance and EOL dates as well as the endoflife.date site for ease of viewing.